pkg/clusterversion/cv.go

```diff
@@ -41,18 +50,96 @@ func (cvb *clusterVersionClientBuilder) New(c client.Client) ClusterVersion {
 
 // GetClusterVersion gets the ClusterVersion CR
 func (c *clusterVersionClient) GetClusterVersion() (*configv1.ClusterVersion, error) {
-	cvList := &configv1.ClusterVersionList{}
-	err := c.client.List(context.TODO(), cvList)
+	cv := &configv1.ClusterVersion{}
+	err := c.client.Get(context.TODO(), types.NamespacedName{Name: OSD_CV_NAME}, cv)
 	if err != nil {
 		return nil, err
 	}
 
-	// ClusterVersion is a singleton
-	for _, cv := range cvList.Items {
-		return &cv, nil
+	return cv, err
+}
+
+func (c *clusterVersionClient) EnsureDesiredVersion(uc *upgradev1alpha1.UpgradeConfig) (bool, error) {
+	clusterVersion, err := c.GetClusterVersion()
+	if err != nil {
+		return false, err
+	}
+
+	// Move the cluster to the same channel first
+	desired := uc.Spec.Desired
+	if clusterVersion.Spec.Channel != desired.Channel {
+		clusterVersion.Spec.Channel = desired.Channel
+		err = c.client.Update(context.TODO(), clusterVersion)
+		if err != nil {
+			return false, err
+		}
+
+		// Retrieve the updated version
+		clusterVersion, err = c.GetClusterVersion()
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// The CVO may need time sync the version before launching the upgrade
+	updateAvailable := false
+	for _, update := range clusterVersion.Status.AvailableUpdates {
+		if update.Version == desired.Version && update.Image != "" {
+			updateAvailable = true
+		}
+	}
+	if !updateAvailable {
+		return false, nil
+	}
+
+	clusterVersion.Spec.Overrides = []configv1.ComponentOverride{}
+	clusterVersion.Spec.DesiredUpdate = &configv1.Update{Version: uc.Spec.Desired.Version}
+	err = c.client.Update(context.TODO(), clusterVersion)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+type HasDegradedOperatorsResult struct {
+	Degraded []string
+}
+
+func (c *clusterVersionClient) HasDegradedOperators() (*HasDegradedOperatorsResult, error) {
+	operatorList := &configv1.ClusterOperatorList{}
+	err := c.client.List(context.TODO(), operatorList, []client.ListOption{}...)
+	if err != nil {
+		return &HasDegradedOperatorsResult{
+			Degraded: []string{},
+		}, err
+	}
+
+	degradedOperators := []string{}
+	for _, co := range operatorList.Items {
+		for _, condition := range co.Status.Conditions {
+			if (condition.Type == configv1.OperatorDegraded && condition.Status == configv1.ConditionTrue) || (condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionFalse) {
+				degradedOperators = append(degradedOperators, co.Name)
```

Comment:
Tiny one here (I know this is a pre-existing bug, I just only noticed it when testing this PR) - can we add a `break` after this line, otherwise in the case that an operator is both degraded and unavailable it adds an entry into the `degradedOperators` for each failing status condition, and it need only appear once.